src/partners/swapuz.ts

```diff
@@ -0,0 +1,181 @@
+  asUnknown
+} from 'cleaners'
+import fetch from 'node-fetch'
+
+import { PartnerPlugin, PluginParams, PluginResult, StandardTx } from '../types'
+import { datelog } from '../util'
+
+const asSwapuzLogin = asObject({
+  result: asObject({
+    token: asString
+  })
+})
+
+const asSwapuzTx = asObject({
+  createDate: asString,
+  from: asString,
+  to: asString,
+  amountResult: asNumber,
+  amount: asNumber,
+  amountBTC: asNumber,
+  startAmount: asNumber
+})
+
+const asSwapuzRawTx = asObject({
+  id: asNumber,
+  status: asNumber,
+  wTxId: asOptional(asString),
+  dTxId: asOptional(asString)
+})
+
+const asSwapuzResult = asObject({
+  result: asObject({
+    currentPage: asNumber,
+    maxPage: asNumber,
+    result: asArray(asUnknown)
+  }),
+  status: asNumber
+})
+
+const asSwapuzPluginParams = asObject({
+  apiKeys: asObject({ login: asString, password: asString }),
+  settings: asObject({
+    latestIsoDate: asOptional(asString, '1970-01-01T00:00:00.000Z')
+  })
+})
+
+const QUERY_LOOKBACK = 1000 * 60 * 60 * 24 * 5 // 5 days
+
+export const querySwapuz = async (
+  pluginParams: PluginParams
+): Promise<PluginResult> => {
+  const ssFormatTxs: StandardTx[] = []
+
+  const { settings, apiKeys } = asSwapuzPluginParams(pluginParams)
+  const { login, password } = apiKeys
+
+  // Get our Bearer token
+  const response = await fetch('https://api.swapuz.com/api/User/login', {
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/json'
+    },
+    body: JSON.stringify({
+      login,
+      password
+    })
+  })
+  if (response.ok === false) {
+    const text = await response.text()
+    throw new Error(text)
+  }
+  const reply = await response.json()
+  const loginReply = asSwapuzLogin(reply)
+
+  const bearerToken = loginReply.result.token
+
+  const headers = {
+    'Content-Type': 'application/json',
+    Authorization: `Bearer ${bearerToken}`
+  }
+
+  let { latestIsoDate } = settings
+
+  let previousTimestamp = new Date(latestIsoDate).getTime() - QUERY_LOOKBACK
+  if (previousTimestamp < 0) previousTimestamp = 0
+  const previousLatestIsoDate = new Date(previousTimestamp).toISOString()
+
+  let page = -1
+  let done = false
+  let oldestIsoDate = new Date(99999999999999).toISOString()
+
+  while (!done) {
+    page++
+    const url = `https://api.swapuz.com/api/partner/partnerPaginator?page=${page}`
+    const response = await fetch(url, {
+      method: 'GET',
+      headers
+    })
+    if (response.ok === false) {
+      const text = await response.text()
+      throw new Error(text)
+    }
+    const reply = await response.json()
+    const jsonObj = asSwapuzResult(reply)
+    const { currentPage, maxPage, result: txs } = jsonObj.result
+    for (const rawTx of txs) {
+      const { id, status, dTxId, wTxId } = asSwapuzRawTx(rawTx)
+
+      // Status === 6 seems to be the "complete" status
+      if (status === 6) {
+        if (dTxId == null) {
+          // console.log(`Status 6 with null dTxId id=${id}`)
```

Comment:
Can we keep these console logs enabled under a flag? I know pino has the ability to enable/disable log levels.

Reply:
These can be removed – Paul.